mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/jobcluster/JobClusterActor.java

```diff
@@ -2162,6 +2223,71 @@ private void updateAndSaveJobCluster(IJobClusterMetadata jobCluster) throws Exce
+    private CompletionStage<Optional<JobDefinition>> getNonTerminalJobDefnWithWorkerCountInheritance(
+            final JobInfo runningJobInfo, final JobDefinition givenJobDefn, boolean forceInheritance) {
+        // Get job details with job metadata tracked inside job actor.
+        logger.info("[Inherit worker count] apply worker count inheritance on given job: {} from {}",
+                givenJobDefn.getName(), runningJobInfo.jobId);
+
+        GetJobDetailsRequest req = new GetJobDetailsRequest("system", runningJobInfo.jobId);
+        return ask(runningJobInfo.jobActor, req, Duration.ofMillis(500))
+                .thenApply(GetJobDetailsResponse.class::cast)
+                .thenApply(res -> {
+                    if (!res.getJobMetadata().isPresent()) {
+                        logger.warn("Got empty job actor metadata.");
+                        return empty();
+                    }
+                    else if (givenJobDefn.getSchedulingInfo() == null) {
+                        logger.warn("Invalid job definition from given job info.");
+                        return empty();
+                    }
+                    else {
+                        JobDefinition.Builder jobDefnBuilder = new JobDefinition.Builder().from(givenJobDefn);
+                        Map<Integer, StageSchedulingInfo> givenStages = givenJobDefn.getSchedulingInfo().getStages();
+
+                        IMantisJobMetadata lastJobMeta = res.getJobMetadata().get();
+                        SchedulingInfo.Builder siBuilder = new SchedulingInfo.Builder().numberOfStages(givenStages.size());
+
+                        givenStages.keySet().stream().sorted().forEach(k -> {
+                            if (lastJobMeta.getStageMetadata(k).isPresent() &&
+                                    (forceInheritance || givenStages.get(k).getInheritInstanceCount())) {
+                                StageSchedulingInfo mergedStageInfo = new StageSchedulingInfo.Builder()
+                                        .setNumberOfInstances(lastJobMeta.getStageMetadata(k).get().getNumWorkers())
+                                        .cloneWithoutNumberOfInstances(givenStages.get(k))
+                                        .build();
+                                if (k == 0) { // handle JobMaster stage
+                                    siBuilder.addJobMasterStage(mergedStageInfo);
+                                }
+                                else { siBuilder.addStage(mergedStageInfo); }
+                            }
+                            else {
+                                if (k == 0) { // handle JobMaster stage
+                                    siBuilder.addJobMasterStage(givenStages.get(k));
+                                }
+                                else { siBuilder.addStage(givenStages.get(k)); }
+                            }
+                        });
+
+                        jobDefnBuilder.withSchedulingInfo(siBuilder.build());
```

Comment:
I'm guessing this block deals with merging the provided job definition and the job definition of the currently running job. I would move this logic to JobDefinition::merge or JobDefinition::concat. This would also allow you to easily unit test this logic.

Reply:
I feel that IMantisJobMetadata logically should not be exposed to the public JobDefinition class. I will convert it to a plain map then.